miniapps/blocksolvers/div_free_solver.hpp

```diff
@@ -0,0 +1,252 @@
+};
+
+/// Block diagonal solver for A, each block is inverted by direct solver
+class BlockDiagSolver : public Solver
+{
+   mutable SparseMatrix block_dof_;
+   mutable Array<int> local_dofs_;
+   mutable Vector sub_rhs_;
+   mutable Vector sub_sol_;
+   Array<DenseMatrixInverse> block_solver_;
+public:
+   BlockDiagSolver(const OperatorPtr& A, SparseMatrix block_dof);
+   virtual void Mult(const Vector &x, Vector &y) const;
+   virtual void MultTranspose(const Vector &x, Vector &y) const { Mult(x, y); }
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// Solver for local problems in SchwarzSmoother
+class LocalSolver : public Solver
+{
+   DenseMatrix local_system_;
+   DenseMatrixInverse local_solver_;
+   const int offset_;
+public:
+   LocalSolver(const DenseMatrix &M, const DenseMatrix &B);
+   virtual void Mult(const Vector &x, Vector &y) const;
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// non-overlapping additive Schwarz for saddle point problems
+class SchwarzSmoother : public Solver
+{
+   const SparseMatrix& agg_hdivdof_;
+   const SparseMatrix& agg_l2dof_;
+   OperatorPtr coarse_l2_projector_;
+
+   Array<int> offsets_;
+   mutable Array<int> offsets_loc_;
+   mutable Array<int> hdivdofs_loc_;
+   mutable Array<int> l2dofs_loc_;
+   Array<OperatorPtr> solvers_loc_;
+public:
+   SchwarzSmoother(const HypreParMatrix& M,
+                   const HypreParMatrix& B,
+                   const SparseMatrix& agg_hdivdof,
+                   const SparseMatrix& agg_l2dof,
+                   const HypreParMatrix& P_l2,
+                   const HypreParMatrix& Q_l2);
+   virtual void Mult(const Vector & x, Vector & y) const;
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// Relaxation on the kernel of divergence
+class KernelSmoother : public Solver
+{
+   Array<int> offsets_;
+   OperatorPtr blk_kernel_map_;
+   OperatorPtr kernel_system_;
+   OperatorPtr kernel_smoother_;
+public:
+   KernelSmoother(const BlockOperator& op, const HypreParMatrix& kernel_map_);
+   virtual void Mult(const Vector & x, Vector & y) const;
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// Solver S such that I - A * S = (I - A * S1) * (I - A * S0).
+/// A, S0, S1 are assumed to be symmetric.
+class ProductSolver : public Solver
+{
+   OperatorPtr op_;
+   Array<OperatorPtr> solvers_;
+   void Mult(int i, int j, const Vector & x, Vector & y) const;
+public:
+   ProductSolver(Operator* A, Operator* S0, Operator* S1,
+                 bool ownA, bool ownS0, bool ownS1)
+      : Solver(A->NumRows()), op_(A, ownA), solvers_(2)
+   {
+      solvers_[0].Reset(S0, ownS0);
+      solvers_[1].Reset(S1, ownS1);
+   }
+   virtual void Mult(const Vector & x, Vector & y) const { Mult(0, 1, x, y); }
+   virtual void MultTranspose(const Vector & x, Vector & y) const { Mult(1, 0, x, y); }
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// Divergence free solver, for more details see Appendix F.3 in
```

Comment:
I don't expect you to write the whole paper in this comment block, but I think this solver deserves a bit more explanation, maybe just the basic idea of what it does.